flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/OperatorCoordinator.java

```diff
@@ -137,7 +145,7 @@ default void notifyCheckpointAborted(long checkpointId) {}
 	 * complete (for example when a system failure happened directly after committing the checkpoint,
 	 * before calling the {@link #notifyCheckpointComplete(long)} method).
 	 */
-	void resetToCheckpoint(byte[] checkpointData) throws Exception;
+	void resetToCheckpoint(@Nullable byte[] checkpointData) throws Exception;
```

Comment:
nit: I think `@Nullable` is a bit more error-prone and less readable than two explicit methods: `reset()` and `resetAndRestore(byte[])`. The latter could call the former by default.

Reply:
For a public API, I would go with two methods. For this internal API, I would not change it at this point.
For the public API of the source, we translate the restore anyways to a re-instantiation.
  - Restore with null state becomes `source.createEnumerator()`
  - Restore with proper state becomes `source.restoreEnumerator()`.

So for users, this confusion should not exist.

Reply:
We also plan to add the `checkpointID` to the call in the future, so the message should be that this is a reset to a checkpoint. Possibly with state.